### PR TITLE
IGAPP-526: Implement jpal settings screen

### PR DESCRIPTION
**Pull request by SteffiStuffel**
This pull request belongs to an issue on our [bugtracker](https://issues.integreat-app.de/).
You can find it there by looking for an issue with the key which is mentioned in the title of this pull request.
It starts with the keyword **IGAPP**.

### CODE REVIEW

File: native/src/routes/jpal-tracking/components/JpalTracking.js [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+  useEffect(() => {
+    const loadSettings = async () => {
+      try {
+        const loadedSettings = await appSettings.loadSettings()
+        setSettingsLoaded(true)
+        setSettingsUpdated(false)
+        setSettings(loadedSettings)
+        setDisplayedTrackingCode(loadedSettings.jpalTrackingCode)
+      } catch (e) {
+        console.error('Failed to load settings.')
```
Comment (maxammann) [inline native/src/routes/jpal-tracking/components/JpalTracking.js:69]: Use an error state variable here to show the user that an error happened.

File: native/src/routes/jpal-tracking/containers/JpalTrackingContainer.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
+// @flow
+
+import type { TFunction } from 'react-i18next'
+import { withTranslation } from 'react-i18next'
+import withTheme from '../../../modules/theme/hocs/withTheme'
+import JpalTracking, { type PropsType as JpalTrackingPropsType } from '../components/JpalTracking'
+
+export default withTheme<$Diff<JpalTrackingPropsType, {| t: TFunction |}>>(
+  withTranslation<JpalTrackingPropsType>('settings')(JpalTracking)
+)
```
Comment (maxammann) [inline native/src/routes/jpal-tracking/containers/JpalTrackingContainer.js:10]: I'm white sure you could use here hooks of react-i18n and our theme.
That would make the tracking screen a container, but I thinkg this is fine.

File: native/src/routes/settings/components/Settings.js [hunk 3]
```diff
@@ -78,6 +83,10 @@ export default class Settings extends React.Component<PropsType, StateType> {
     }
   }
 
+  componentWillUnmount() {
+    this.unsubscribeNavigationListener()
+  }
```
Comment (maxammann) [inline native/src/routes/settings/components/Settings.js:88]: Why is this needed? why do we need to load settings when the comonent unmounts?

**Review by maxammann — APPROVED** [review]

File: version.json [hunk 4]
```diff
@@ -1 +1,4 @@
-{"versionName":"2021.3.4","versionCode":100042155}
\ No newline at end of file
+{
+  "versionName": "2021.3.4",
+  "versionCode": 100042155
+}
```
Comment (klinzo) [inline version.json:4]: Revert this change please, probably because of prettier, right? I added it to ignored files somewhere already, not sure in which PR though.

File: translations/translations.json [hunk 5]
```diff
@@ -2994,6 +2994,8 @@
       "version": "Version: {{version}}",
       "openSourceLicenses": "Open-Source-Lizenzen",
       "tracking": "J-PAL Evaluation",
+      "trackingAllowed": "Tracking zulassen",
```
Comment (klinzo) [inline translations/translations.json:2997]:
```suggestion
      "allowTracking": "Tracking zulassen",
```

File: native/src/routes/jpal-tracking/components/JpalTracking.js [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+export type PropsType = {|
+  theme: ThemeType,
+  t: TFunction,
+  route: RoutePropType<JpalTrackingRouteType>,
+  navigation: NavigationPropType<JpalTrackingRouteType>
+|}
+
+const JpalTracking = (props: PropsType, state: StateType) => {
+  const [appSettings] = useState(new AppSettings())
```
Comment (klinzo) [inline native/src/routes/jpal-tracking/components/JpalTracking.js:47]: This seems weird, I think keeping the settings in the state should be enough. Although I think you should type `useState<type>`.

File: native/src/routes/jpal-tracking/components/JpalTracking.js [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+        setSettings(loadedSettings)
+        setDisplayedTrackingCode(loadedSettings.jpalTrackingCode)
+      } catch (e) {
+        console.error('Failed to load settings.')
+      }
+    }
+    loadSettings()
+  }, [appSettings, settingsUpdated])
+
+  const onPress = () => {
```
Comment (klinzo) [inline native/src/routes/jpal-tracking/components/JpalTracking.js:75]: I think a more speaking name like e.g. `toggleTrackingEnabled` would make sense if this is what it does.

File: native/src/routes/jpal-tracking/components/JpalTracking.js [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+        setSettingsLoaded(true)
+        setSettingsUpdated(false)
+        setSettings(loadedSettings)
+        setDisplayedTrackingCode(loadedSettings.jpalTrackingCode)
+      } catch (e) {
+        console.error('Failed to load settings.')
+      }
+    }
+    loadSettings()
+  }, [appSettings, settingsUpdated])
```
Comment (klinzo) [inline native/src/routes/jpal-tracking/components/JpalTracking.js:73]: I don't think we have to reload the settings every time we change something. I think it is enough to just load the settings once initially and keep only relevant settings in a local state. Then you could just catch possible errors during setting of the tracking code or tracking enabled to the global app settings and change the local state if everything succeeds and no error is thrown. Than you can safely assume that the local state always matches the global state.

File: native/src/routes/jpal-tracking/components/JpalTracking.js (hunk 1 above)
Comment (klinzo) [reply native/src/routes/jpal-tracking/components/JpalTracking.js:69]: Possibly also just with a todo to add a snackbar at a later point of time.

File: native/src/routes/jpal-tracking/components/__tests__/JpalTracking.spec.js [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+  const route: RoutePropType<JpalTrackingRouteType> = {
+    key: 'route-id-0',
+    params: { trackingCode: 'abc' },
+    name: JPAL_TRACKING_ROUTE
+  }
+
+  it('Textfield should not be editable, when tracking is disabled', async () => {
+    const { findByTestId } = render(<JpalTracking theme={lightTheme} t={t} route={route} navigation={navigation} />)
+
+    AsyncStorage.setItem('jpalTrackingEnabled', true, async () => {
```
Comment (klinzo) [inline native/src/routes/jpal-tracking/components/__tests__/JpalTracking.spec.js:31]: I'd rather use our AppSettings instead of AsyncStorage directly. I also think it is weird to render the component before you set the setting initially. Maybe it also makes sense to assert on a few more things like that jpal tracking is enabled in the AppSettings if you toggled the switch.

File: native/src/routes/jpal-tracking/components/JpalTracking.js [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+  const { jpalTrackingEnabled } = settings
+
+  return (
+    <View style={{ padding: 40 }}>
+      <Caption title={t('tracking')} theme={theme} />
+      <Text>{t('trackingDescription')}</Text>
+
+      <DescriptionContainer>
+        <ThemedText theme={props.theme}>{t('trackingAllowed')}</ThemedText>
+        <Switch
```
Comment (klinzo) [inline native/src/routes/jpal-tracking/components/JpalTracking.js:99]: Maybe it makes sense to switch on a click on the text as well? I.e. wrap the text in the switch.

**Review by klinzo — CHANGES_REQUESTED** [review]

File: native/src/routes/settings/components/Settings.js (hunk 3 above)
Comment (SteffiStuffel) [reply native/src/routes/settings/components/Settings.js:88]: In settings screen we see if tracking is enabled or disabled, when clicking on tracking-settings and navigate to jpal-settings-screen, we can update if tracking is enabled and when nativate.goBack to general settings, there is still the old value, because settings are only loaded in constructor. Therefore i add an listener for "on-focus" so component loads settings when opend (also on goBack()). We have to unsubscribe from this listener when unmounting.

File: native/src/routes/jpal-tracking/components/JpalTracking.js (hunk 1 above)
Comment (SteffiStuffel) [reply native/src/routes/jpal-tracking/components/JpalTracking.js:69]: done

File: translations/translations.json (hunk 5 above)
Comment (SteffiStuffel) [reply translations/translations.json:2997]: done

File: native/src/routes/jpal-tracking/components/JpalTracking.js (hunk 7 above)
Comment (SteffiStuffel) [reply native/src/routes/jpal-tracking/components/JpalTracking.js:75]: done

File: native/src/routes/jpal-tracking/components/JpalTracking.js (hunk 10 above)
Comment (SteffiStuffel) [reply native/src/routes/jpal-tracking/components/JpalTracking.js:99]: So clicking on the text field would be a consent for tracking? Not sure if this is a good idea.

Comment (klinzo) [reply native/src/routes/jpal-tracking/components/JpalTracking.js:99]: I think that is the default behavior in apps, clicking somewhere in the row always toggles settings. Maybe we need a dialog at a later time depending on what texts, terms and conditions jpal gives us.

File: native/src/routes/jpal-tracking/components/JpalTracking.js (hunk 8 above)
Comment (SteffiStuffel) [reply native/src/routes/jpal-tracking/components/JpalTracking.js:73]: done

File: native/src/routes/jpal-tracking/components/JpalTracking.js (hunk 6 above)
Comment (SteffiStuffel) [reply native/src/routes/jpal-tracking/components/JpalTracking.js:47]: if just putting it like:

const appSettings = new AppSettings() it get:

ESLint: The 'appSettings' object construction makes the dependencies of useEffect Hook (at line 73) change on every render. To fix this, wrap the initialization of 'appSettings' in its own useMemo() Hook.(react-hooks/exhaustive-deps)

i now changed it to useMemo()

File: native/src/routes/jpal-tracking/containers/JpalTrackingContainer.js (hunk 2 above)
Comment (SteffiStuffel) [reply native/src/routes/jpal-tracking/containers/JpalTrackingContainer.js:10]: changed it for translations, for theme we would need a ThemeProvider, we do not have yet, as Steffen poineted out.

File: native/src/routes/jpal-tracking/components/JpalTracking.js (hunk 10 above)
Comment (SteffiStuffel) [reply native/src/routes/jpal-tracking/components/JpalTracking.js:99]: I think we talk about different things. 
But now i get what you mean: The whole row and not only the switch should be clickable?

File: native/src/routes/jpal-tracking/__tests__/JpalTracking.spec.js [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+  const route: RoutePropType<JpalTrackingRouteType> = {
+    key: 'route-id-0',
+    params: { trackingCode: 'abc' },
+    name: JPAL_TRACKING_ROUTE
+  }
+
+  // Error: Unable to find node on an unmounted component.
+  // eslint-disable-next-line jest/no-disabled-tests
+  it.skip('Textfield should not be editable, when tracking is disabled', async () => {
+    await act(async () => {
```
Comment (SteffiStuffel) [inline native/src/routes/jpal-tracking/__tests__/JpalTracking.spec.js:35]: Unfortunately I was not able to get this test running, need some help here.

File: native/src/routes/jpal-tracking/JpalTracking.js [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+    })
+  }
+
+  const setTrackingCode = (value: string) => {
+    setDisplayedTrackingCode(value)
+    appSettings.setJpalTrackingCode(value).catch(() => {
+      setError(true)
+      setSettingsUpdated(true)
+    })
+    props.navigation.setParams({ trackingCode: value })
```
Comment (klinzo) [inline native/src/routes/jpal-tracking/JpalTracking.js:103]: I think it is enough to just set the tracking code from the navigation params initially. No need to keep the tracking code in the navigation params consistent afterwards.

File: native/src/routes/jpal-tracking/JpalTracking.js [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+      }
+    }
+    loadSettings()
+  }, [appSettings, settingsUpdated])
+
+  const toggleTrackingEnabled = () => {
+    setDisplayedTrackingEnabled(!displayedTrackingEnabled)
+    appSettings.setJpalTrackingEnabled(!settings.jpalTrackingEnabled).catch(() => {
+      setError(true)
+      setSettingsUpdated(true)
```
Comment (klinzo) [inline native/src/routes/jpal-tracking/JpalTracking.js:93]: Did you try to get it working without this `settingsUpdated` thingy?

File: native/src/routes/jpal-tracking/JpalTracking.js [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+
+export type PropsType = {|
+  theme: ThemeType,
+  route: RoutePropType<JpalTrackingRouteType>,
+  navigation: NavigationPropType<JpalTrackingRouteType>
+|}
+
+const errorDisplayTime = 5000
+
+const JpalTracking = (props: PropsType, state: StateType) => {
```
Comment (klinzo) [inline native/src/routes/jpal-tracking/JpalTracking.js:54]:
```suggestion
const JpalTracking = (props: PropsType) => {
```

File: native/src/routes/jpal-tracking/JpalTracking.js [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+import styled from 'styled-components/native'
+import LayoutContainer from '../../modules/layout/containers/LayoutContainer'
+import withTheme from '../../modules/theme/hocs/withTheme'
+import Caption from '../../modules/common/components/Caption'
+
+const ThemedText = styled.Text`
+  display: flex;
+  text-align: left;
+  color: ${props => props.theme.colors.textColor};
+  font-family: ${props => props.theme.fonts.decorativeFontRegular};
```
Comment (klinzo) [inline native/src/routes/jpal-tracking/JpalTracking.js:21]: Otherwise the app crashes:
```suggestion
  font-family: ${props => props.theme.fonts.native.decorativeFontRegular};
```

File: native/src/routes/jpal-tracking/JpalTracking.js [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+  }
+
+  if (!settingsLoaded) {
+    return <LayoutContainer />
+  }
+
+  const { theme } = props
+
+  return (
+    <View style={{ padding: 40 }}>
```
Comment (klinzo) [inline native/src/routes/jpal-tracking/JpalTracking.js:113]: Should be wrapped in a LayoutContainer to ensure the right background color and layout :)

File: native/src/routes/jpal-tracking/JpalTracking.js [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+  const [displayedTrackingCode, setDisplayedTrackingCode] = useState('')
+  const [displayedTrackingEnabled, setDisplayedTrackingEnabled] = useState(false)
+  const [error, setError] = useState(false)
+  const routeTrackingCode = props.route.params.trackingCode
+
+  const { t } = useTranslation(['settings', 'error'])
+
+  useEffect(() => {
+    if (routeTrackingCode) {
+      appSettings.setJpalTrackingCode(routeTrackingCode)
```
Comment (klinzo) [inline native/src/routes/jpal-tracking/JpalTracking.js:68]: Shouldn't we display the preset tracking code then? What happens if the route params contains a code different from one already saved in the settings?

File: native/src/routes/jpal-tracking/components/JpalTracking.js (hunk 6 above)
Comment (klinzo) [reply native/src/routes/jpal-tracking/components/JpalTracking.js]: You could also just move it out of the component, i.e.:
```
const appSettings = new AppSettings()

const JpalTracking = (props: PropsType) => {
...
```
Then you also don't have to put it into the dependency of the hooks.

File: native/src/routes/jpal-tracking/JpalTracking.js (hunk 13 above)
Comment (SteffiStuffel) [reply native/src/routes/jpal-tracking/JpalTracking.js:93]: yes works now

File: native/src/routes/jpal-tracking/JpalTracking.js (hunk 17 above)
Comment (SteffiStuffel) [reply native/src/routes/jpal-tracking/JpalTracking.js:68]: It gets changed. Thought this would make sense. If one enables tracking and enters a code, and afterwards gets an letter from jpal an opens a deeplink with tracking code, i would expect it to update.

Comment (klinzo) [reply native/src/routes/jpal-tracking/JpalTracking.js:67]: It gets changed through loading it from the global app settings afterwards, right? So it is a little race condition, correct?

Yeah, probably you are right as there is no use case where a user clicks a jpal link with tracking code and does not want to change the tracking code. Other opinions @maxammann @Taggotty @svenseeberg? Do you think it is a problem to overwrite a previously set tracking code if a user clicks on a link containing a different tracking code?

File: native/src/routes/jpal-tracking/JpalTracking.js [hunk 18]
```diff
@@ -0,0 +1,138 @@
...
+        setError(true)
+      }
+    }
+    loadSettings()
+  }, [])
+
+  const toggleTrackingEnabled = () => {
+    setDisplayedTrackingEnabled(!displayedTrackingEnabled)
+    appSettings.setJpalTrackingEnabled(!settings.jpalTrackingEnabled).catch(() => {
+      setError(true)
```
Comment (klinzo) [inline native/src/routes/jpal-tracking/JpalTracking.js:90]: Should we reset the displayed/local settings to the value it was previously if an error happens here?

File: native/src/routes/jpal-tracking/JpalTracking.js [hunk 19]
```diff
@@ -0,0 +1,138 @@
...
+  navigation: NavigationPropType<JpalTrackingRouteType>
+|}
+
+const errorDisplayTime = 5000
+
+const appSettings = new AppSettings()
+
+const JpalTracking = (props: PropsType) => {
+  const [settings, setSettings] = useState(defaultSettings)
+  const [settingsLoaded, setSettingsLoaded] = useState(false)
```
Comment (klinzo) [inline native/src/routes/jpal-tracking/JpalTracking.js:57]: As mentioned in the coworking: I think those could be omitted :)

File: native/src/routes/jpal-tracking/JpalTracking.js [hunk 20]
```diff
@@ -0,0 +1,138 @@
...
+          />
+        </DescriptionContainer>
+
+        <ThemedText theme={props.theme}>{t('trackingCode')}</ThemedText>
+        <Input
+          value={displayedTrackingCode}
+          onChangeText={setTrackingCode}
+          theme={theme}
+          editable={displayedTrackingEnabled}
+          testID='input'></Input>
```
Comment (klinzo) [inline native/src/routes/jpal-tracking/JpalTracking.js:132]:
```suggestion
        <Input
          value={displayedTrackingCode}
          onChangeText={setTrackingCode}
          theme={theme}
          editable={displayedTrackingEnabled}
          testID='input' />
```

Comment (klinzo) [reply native/src/routes/jpal-tracking/JpalTracking.js:132]: Could I have a grey border if it is not editable please? :stuck_out_tongue:

**Review by klinzo — APPROVED** [review]

File: native/src/routes/jpal-tracking/JpalTracking.js [hunk 21]
```diff
@@ -0,0 +1,136 @@
...
+    }
+  }, [routeTrackingCode])
+
+  useEffect(() => {
+    const loadSettings = async () => {
+      try {
+        const loadedSettings = await appSettings.loadSettings()
+        setSettingsLoaded(true)
+        setSettings(loadedSettings)
+        setTimeout(() => setError(false), errorDisplayTime)
```
Comment (klinzo) [inline native/src/routes/jpal-tracking/JpalTracking.js:75]: Use effect should return a function removing all timeouts and pending promises

Comment (klinzo) [reply native/src/routes/jpal-tracking/JpalTracking.js:75]: Also, why do we have a timeout for the error just here?

File: native/src/routes/jpal-tracking/JpalTracking.js (hunk 17 above)
Comment (SteffiStuffel) [reply native/src/routes/jpal-tracking/JpalTracking.js:65]: removed the race condition.

File: native/src/routes/jpal-tracking/JpalTracking.js (hunk 18 above)
Comment (SteffiStuffel) [reply native/src/routes/jpal-tracking/JpalTracking.js:86]: done

File: native/src/routes/jpal-tracking/JpalTracking.js (hunk 19 above)
Comment (SteffiStuffel) [reply native/src/routes/jpal-tracking/JpalTracking.js:57]: We talked abour settingsUpdated, i removed it. But i think we should keep settingsLoaded, otherwise the page is rendered, when the date is not present yet and then (after a short time), when the data is present, the switch and the textfield get updated, what looks weird

File: native/src/routes/jpal-tracking/JpalTracking.js (hunk 20 above)
Comment (SteffiStuffel) [reply native/src/routes/jpal-tracking/JpalTracking.js]: It is there already :)

File: native/src/routes/jpal-tracking/JpalTracking.js (hunk 21 above)
Comment (SteffiStuffel) [reply native/src/routes/jpal-tracking/JpalTracking.js:75]: cause it is the timeout to "unset" the error, so the error is displayed until the settings are reloaded correctly plus errorDisplayTime (3s). Think it does not make sens anywhere else.

**Comment by maxammann** [comment]
> Do you think it is a problem to overwrite a previously set tracking code if a user clicks on a link containing a different tracking code?

I dont think that is a problem.

**Review by klinzo — APPROVED** [review]
